### PR TITLE
Use EKG to monitor, custom nt, nt-tcp, tw-nt

### PR DESCRIPTION
**Pull request by georgeee**
The custom network-transport, network-transport-tcp, and time-warp-ntbranches chosen in stack.yaml fix some potentially very bad problemswith bidirectional connections: a handler can hang indefinitely in casethe connection is lost to the peer before the ACK is received. Butthat's fixed now. The nt-tcp branch is also patched to deliverConnectionClosed events always, even if the connection is lost of theendpoint is brought down. This makes it easier to clean up inputchannels. The dispatcher remains defensive, though, and will plug inputchannels on end point closed and connection lost events, even though itshouldn't ever have to.

### CODE REVIEW

File: src/Pos/Launcher/Runner.hs [hunk 1]
```diff
@@ -185,15 +193,38 @@ runRawRealMode res np@NodeParams {..} sscnp listeners outSpecs (ActionSpec actio
        runDBHolder modernDBs . runCH np initNC $ initNodeDBs
        initTip <- runDBHolder modernDBs getTip
        stateM <- liftIO SM.newIO
+       stateM_ <- liftIO SM.newIO
+
+       -- TODO need an effect-free way of running this into IO.
+       let runIO :: forall t . RawRealMode ssc t -> IO t
+           runIO = runProduction .
```
Comment (georgeee) [inline src/Pos/Launcher/Runner.hs]: @gromakovsky please check this function

File: src/Pos/Launcher/Runner.hs [hunk 2]
```diff
@@ -211,24 +242,44 @@ runServiceMode res bp@BaseParams{..} listeners outSpecs (ActionSpec action) = do
...
     let (listeners', InSpecs ins, OutSpecs outs) = unpackLSpecs packedLS
         ourVerInfo = VerInfo protocolMagic lastKnownBlockVersion ins $ outs <> wouts
         listeners = listeners' ourVerInfo ++ protocolListeners
     stdGen <- liftIO newStdGen
     logInfo $ sformat ("Our verInfo "%build) ourVerInfo
     node (concrete transport) stdGen BiP (ourPeerId, ourVerInfo) $ \__node ->
-        pure $ NodeAction listeners (action ourVerInfo)
+        pure $ NodeAction listeners $ \sendActions -> do
+            st <- liftIO getMaskingState
+            logInfo $ sformat ("Node masking state is " % shown) st
```
Comment (georgeee) [inline src/Pos/Launcher/Runner.hs]: What is masking state, @avieth ?

File: src/Pos/Ssc/Extra/Holder.hs [hunk 3]
```diff
@@ -88,19 +90,47 @@ instance Monad m => MonadSscMem ssc (SscHolder ssc m) where
...
+       ( WithLogger m
+       , WithNodeContext ssc m
+       , SscGStateClass ssc
+       , SscLocalDataClass ssc
+       , MonadDB ssc m
+       , MonadSlots m
+       )
+    => SscHolder ssc m a
+    -> m a
+mkStateAndRunSscHolder holder = do
```
Comment (georgeee) [inline src/Pos/Ssc/Extra/Holder.hs]: @gromakovsky here as well

File: src/Pos/Launcher/Runner.hs (hunk 2 above)
Comment (avieth) [inline src/Pos/Launcher/Runner.hs]: It's the async exception masking state. I'll remove this. Just had it here to verify that it's `Unmasked` at this point.

File: src/Pos/Launcher/Runner.hs (hunk 1 above)
Comment (avieth) [inline src/Pos/Launcher/Runner.hs]: About this: to set up EKG, we need to be able to run the special monad `m`, in this state `RawRealMode`, into `IO`, i.e. we need `forall t . RawRealMode t -> IO t`. BUT we need to do this in an 'effect-free' way. Unfortunately, some of the 'runner' functions, `runSscHolder` in particular, have extra side-effects. Ideally we would do the ssc runner side-effects once and use the resulting values to obtain `SscHolder m t -> m t` which can we re-used for this `runIO` and also for the function below which runs the node into `Production`.

Comment (gromakovsky) [inline src/Pos/Launcher/Runner.hs]: Everything looks ok.
Regarding described problem: yes, I see. I think we will do this refactoring after we unite all these holders into one holder (which is another refactoring).

**Pull request merged**